system/security/plugins/jwtSecurity/jwtEndpoint.cpp

```diff
@@ -0,0 +1,192 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include <curl/curl.h>
+#include <iostream>
+#include <string.h>
+
+#include "jlog.hpp"
+
+#include "nlohmann/json.hpp"
+
+#include "jwtEndpoint.hpp"
+
+struct MemoryStruct
+{
+    char*   memory;
+    size_t  size;
+    size_t  capacity;
+};
+
+void InitMemoryStructCapacity(MemoryStruct& mem, size_t initCapacity)
+{
+    mem.memory = static_cast<char*>(malloc(initCapacity));
+    mem.size = 0;
+    mem.capacity = initCapacity;
+}
+
+void FreeMemoryStruct(MemoryStruct& mem)
+{
+    if (mem.memory)
+    {
+        free(mem.memory);
+        mem.memory = nullptr;
+    }
+    mem.size = 0;
+    mem.capacity = 0;
+}
+
+size_t CaptureIncomingCURLReply(void* contents, size_t size, size_t nmemb, void* userp)
+{
+    size_t          incomingDataSize = size * nmemb;
+    MemoryStruct*   mem = static_cast<struct MemoryStruct*>(userp);
+    size_t          MAX_BUFFER_SIZE = 4194304; // 2^22
+
+    if ((mem->size + incomingDataSize) < MAX_BUFFER_SIZE)
+    {
+        if ((mem->size + incomingDataSize) > mem->capacity)
+        {
+            size_t  newCapacity = mem->capacity * 2;
+
+            // Keep doubling capacity until it is greater than what we need
+            while ((mem->size + incomingDataSize) > newCapacity)
+                newCapacity *= 2;
+
+            mem->memory = static_cast<char*>(realloc(mem->memory, newCapacity));
+            mem->capacity = newCapacity;
+        }
+
+        memcpy(&(mem->memory[mem->size]), contents, incomingDataSize);
+        mem->size += incomingDataSize;
+    }
+    else
+    {
+        // Signals an error to libcurl
+        incomingDataSize = 0;
+
+        PROGLOG("CJwtSecurityManager::CaptureIncomingCURLReply() exceeded buffer size %ld", MAX_BUFFER_SIZE);
+    }
+
+    return incomingDataSize;
+}
+
+std::string _tokenFromEndpoint(const std::string& jwtEndPoint, bool allowSelfSignedCert, const std::string& credentialsStr)
+{
+    std::string     apiResponse;
+
+    // Call JWT endpoint
+    CURL*   curlHandle = curl_easy_init();
```

Comment:
Have you thought through the thread safety if this is the place the global curl init gets done?  What if we start using libcurl more places?  It looks like our current use of libcurl all comes from 3rd party libraries... but once you start using it other components might follow suit.

Reply:
"If you did not already call curl_global_init, curl_easy_init does it automatically. This may be lethal in multi-threaded cases, since curl_global_init is not thread-safe, and it may result in resource problems because there is no corresponding cleanup.

You are strongly advised to not allow this automatic behaviour, by calling curl_global_init yourself properly. See the description in libcurl(3) of global environment requirements for details of how to use this function."

Reply:
FYI I recently added https://github.com/hpcc-systems/HPCC-Platform/tree/master/system/httplib for REST calls related to my secrets.  I didn't do a plus and minus comparison with libcurl.  I wasn't aware it was already a build dependency.  httplib is a single header implementation which simplified the dependencies (and perhaps other issues).

Reply:
httplib looks interesting.  It is definitely a subset implementation, though.  I wonder how it handles redirects and self-signed certs, for instance.  (Those two things come to mind just from this JWT work.)

In looking around for an example of a module-based construction/destruction I stumbled across an azure library in our code base that seems to also initialize libcurl incorrectly:  azure/azure-storage-cpplite/include/http/libcurl_http_client.h.

Would it be appropriate to put the global init management into MODULE_INIT/MODULE_EXIT functions to at least mitigate problems with this library?

Reply:
A proper call to the global init should then fix any use of curl_easy_init() later since it only calls global_init if it hasn't already.  Since it can apply to other 3rd party usage as well we may want a real fix.

@ghalliday any thoughts on where or how we should make a truly threadsafe global call curl_global_init()?  This could also benefit the AWS library use of libcurl.

From AWS library:

        bool initAndCleanupCurl;
        /**
        * libCurl infects everything with its global state. If it is being used then we automatically initialize and clean it up.
        * If this is a problem for you, set this to false. If you manually initialize libcurl please add the option CURL_GLOBAL_ALL to your init call.
        */

Reply:
It would be cool to have the engines perform global libcurl init/cleanup at binary execution/termination time.  From what I see in the docs, that would be safe and ensure safety for any function calling curl_easy_init().

Reply:
@afishbeck I think it would be safe to call it in one of the initialisation blocks.  E.g.

```
MODULE_INIT(INIT_PRIORITY_STANDARD)
{
    cur_global_init();
    return true;
}
MODULE_EXIT()
{
    curl_global_cleanup();
}

```
Then it should be called after the dll has been loaded - even if it is loaded dynamically - which should avoid the potential loaded locking the documentation mentions.

That does mean that any options set by other uses will be ignored.  I'm not sure if that will cause issues.